### PR TITLE
sort resource comparation functions, add accurate definitions and fix…

### PR DESCRIPTION
**Pull request by Thor-wl**
The modifications are as follows:
1. set min resource as const
2. resort functions as groups according its function
3. add assurate definitions for resource comparation functions
4. fix some bugs such as lack of validation for keys in map
5. delete `Scale` function for it is all the same with `Multi`

Signed-off-by: Thor-wl <13164644535@163.com>

### CODE REVIEW

File: pkg/scheduler/api/resource_info.go [hunk 1]
```diff
@@ -208,30 +245,23 @@ func (r *Resource) FitDelta(rr *Resource) *Resource {
...
-			r.ScalarResources = map[v1.ResourceName]float64{}
-		}
+	if r.ScalarResources == nil {
+		r.ScalarResources = make(map[v1.ResourceName]float64)
+	}
 
-		if rrQuant > 0 {
+	for rrName, rrQuant := range rr.ScalarResources {
+		_, ok := r.ScalarResources[rrName]
+		if ok && rrQuant > 0 {
```
Comment (wpeng102) [inline pkg/scheduler/api/resource_info.go:254]: if r does not have `rrName` also need do `r.ScalarResources[rrName] -= rrQuant + minResource`

Comment (Thor-wl) [reply pkg/scheduler/api/resource_info.go:254]: > if r does not have `rrName` also need do `r.ScalarResources[rrName] -= rrQuant + minResource`

If that, we should initially set `r.ScalarResources[rrName] == 0`

**Review by wpeng102 — APPROVED** [review]
/lgtm

File: pkg/scheduler/api/resource_info.go [hunk 2]
```diff
@@ -68,32 +93,44 @@ func (r *Resource) Clone() *Resource {
...
+		return r.ScalarResources[rn]
 	}
-	return r
 }
 
-// IsEmpty returns bool after checking any of resource is less than min possible value
+// IsEmpty returns false if any kind of resource is not less than min value, otherwise returns true
 func (r *Resource) IsEmpty() bool {
-	if !(r.MilliCPU < minResource && r.Memory < minResource) {
+	if r.MilliCPU >= minResource || r.Memory >= minResource {
```
Comment (william-wang) [inline pkg/scheduler/api/resource_info.go:133]: do not use '>=' for float comparing.  Combining '!' and '<'  is a general way.

File: pkg/scheduler/api/resource_info.go [hunk 3]
```diff
@@ -40,17 +49,33 @@ type Resource struct {
...
-)
-
 // EmptyResource creates a empty resource object and returns
 func EmptyResource() *Resource {
 	return &Resource{}
 }
 
-// Clone is used to clone a resource type
+// NewResource creates a new resource object from resource list
+func NewResource(rl v1.ResourceList) *Resource {
```
Comment (william-wang) [inline pkg/scheduler/api/resource_info.go:58]: If there is no change for the function. We'd better keep it as it is.

Comment (Thor-wl) [reply pkg/scheduler/api/resource_info.go:58]: > If there is no change for the function. We'd better keep it as it is.

No changes about these functions, just resort their orders.

File: pkg/scheduler/api/resource_info.go (hunk 2 above)
Comment (Thor-wl) [reply pkg/scheduler/api/resource_info.go:133]: > do not use '>=' for float comparing. Combining '!' and '<' is a general way.

Yes, i will recover it

**Review by william-wang — APPROVED** [review]
/lgtm
/approve

**Review by shinytang6 — APPROVED** [review]
/lgtm
/approve

**Pull request merged**